File: src/cspfj/constraint/Matrix2D.java
package cspfj.constraint;

import cspfj.util.BooleanArray;

public class Matrix2D implements Matrix, Cloneable {
	private int[][] xMatrix;
	private int[][] yMatrix;

	public Matrix2D(int x, int y, boolean initialState) {

		xMatrix = new int[x][BooleanArray.booleanArraySize(y)];
		for (int[] m : xMatrix) {
			BooleanArray.initBooleanArray(m, y, initialState);
		}

		yMatrix = new int[y][BooleanArray.booleanArraySize(x)];

		for (int[] m : yMatrix) {
			BooleanArray.initBooleanArray(m, x, initialState);
		}

	}

	@Override
	public boolean check(final int[] tuple) {
		return BooleanArray.isTrue(xMatrix[tuple[0]], tuple[1]);
	}

	@Override
	public void set(final int[] tuple, final boolean status) {
		BooleanArray.set(xMatrix[tuple[0]], tuple[1], status);
		BooleanArray.set(yMatrix[tuple[1]], tuple[0], status);
	}

	public int[] getBooleanArray(final int position, final int index) {
		switch (position) {
		case 0:
			return xMatrix[index];
		case 1:
			return yMatrix[index];
		default:
			throw new IllegalArgumentException();
		}
	}

	public Matrix2D clone() throws CloneNotSupportedException {
		final Matrix2D matrix2d = (Matrix2D) super.clone();

		final int yLength = xMatrix[0].length;
		matrix2d.xMatrix = new int[xMatrix.length][yLength];
		for (int j = xMatrix.length; --j >= 0;) {
			System.arraycopy(xMatrix[j], 0, matrix2d.xMatrix[j], 0,
					yLength);
		}
		
		final int xLength = yMatrix[0].length;
		matrix2d.yMatrix = new int[yMatrix.length][xLength];
		for (int j = yMatrix.length; --j >= 0;) {
			System.arraycopy(yMatrix[j], 0, matrix2d.yMatrix[j], 0,
					xLength);
		}	

		return matrix2d;

	}

	public String toString() {
		final StringBuilder stb = new StringBuilder();
		final int[] tuple = new int[2];
		for (int i = 0; i < xMatrix.length; i++) {
			tuple[0] = i;
			for (int j = 0; j < yMatrix.length; j++) {
				tuple[1] = j;
				stb.append(check(tuple) ? 1 : 0);
			}
			stb.append('\n');
		}
		return stb.toString();
	}

}

File: src/main/scala/concrete/constraint/extension/Matrix2D.scala
package concrete.constraint.extension

import bitvectors.BitVector
import com.typesafe.scalalogging.LazyLogging

final class Matrix2D(xSize: Int, ySize: Int, val xOffset: Int, val yOffset: Int, initialState: Boolean) extends Matrix with LazyLogging {
  private var xMatrix = {
    val bv = if (initialState) {
      BitVector.filled(ySize)
    } else {
      BitVector.empty
    }
    Array.fill(xSize)(bv)
  }

  private var yMatrix = {
    val bv = if (initialState) {
      BitVector.filled(xSize)
    } else {
      BitVector.empty
    }
    Array.fill(ySize)(bv)
  }

  private var empty = initialState

  def size: Int = xMatrix.length * yMatrix.length

  override def set(tuple: Array[Int], status: Boolean): Unit = {
    val Array(rx, ry) = tuple
    set(rx, ry, status)
  }

  def set(rx: Int, ry: Int, status: Boolean): Unit = {
    val x = rx - xOffset
    val y = ry - yOffset
    require(x >= 0)
    require(y >= 0)
    require(x < xMatrix.length, s"Tuple ($rx, $ry) is out of the scope of matrix $this")
    require(y < yMatrix.length, s"Tuple ($rx, $ry) is out of the scope of matrix $this")
    //if (x < xMatrix.length && y < yMatrix.length) {
    if (status) {
      xMatrix(x) += y
      yMatrix(y) += x
    } else {
      xMatrix(x) -= y
      yMatrix(y) -= x
    }
    empty = false
    //}

    //    else {
    //      logger.info(s"Tuple ($rx, $ry) is out of the scope of matrix $this")
    //    }

  }

  def offsets(position: Int):Int = {
    assert(0 <= position && position < 2)
    if (position == 0) {
      xOffset
    } else {
      yOffset
    }
  }

  def getBitVector(position: Int, index: Int): BitVector = {
    assert(0 <= position && position < 2)
    if (position == 0) {
      xMatrix(index - xOffset)
    } else {
      yMatrix(index - yOffset)
    }
  }

  def copy: Matrix2D = {
    val matrix2d = new Matrix2D(xSize, ySize, xOffset, yOffset, initialState)
    matrix2d.xMatrix = xMatrix.clone
    matrix2d.yMatrix = yMatrix.clone
    matrix2d
  }

  override def toString: String = {
    val stb = new StringBuilder
    stb.append(s"x offset: $xOffset, y offset: $yOffset\n")
    for (i <- xMatrix.indices) {
      for (j <- yMatrix.indices) {
        if (xMatrix(i)(j)) {
          stb.append(1)
        } else {
          stb.append(0)
        }
      }
      stb.append('\n')
    }
    stb.toString
  }

  override def check(tuple: Array[Int]):Boolean = xMatrix(tuple(0) - xOffset)(tuple(1) - yOffset)

  def isEmpty: Boolean = empty

  def allowed: Iterator[Array[Int]] = {
    for {
      x <- (0 until xSize).iterator
      y <- xMatrix(x).iterator
    } yield {
      Array(x + xOffset, y + yOffset)
    }
  }

}
